scan-command/src/main/java/io/ballerina/scan/internal/StaticCodeAnalyzer.java

```diff
@@ -100,6 +120,95 @@ public void visit(ImplicitAnonymousFunctionExpressionNode implicitAnonymousFunct
         this.visitSyntaxNode(implicitAnonymousFunctionExpressionNode.expression());
     }
 
+    @Override
+    public void visit(ClassDefinitionNode classDefinitionNode) {
+        checkNonIsolatedPublicClassDefinition(classDefinitionNode);
+        checkNonIsolatedPublicClassMembers(classDefinitionNode);
+    }
+
+    @Override
+    public void visit(TypeDefinitionNode typeDefinitionNode) {
+        semanticModel.symbol(typeDefinitionNode).ifPresent(symbol -> {
+            if (symbol instanceof TypeDefinitionSymbol typeDefinitionSymbol) {
+                TypeSymbol typeSymbol = typeDefinitionSymbol.typeDescriptor();
+                if (typeSymbol instanceof ObjectTypeSymbol objectTypeSymbol) {
+                    List<Qualifier> qualifiers = objectTypeSymbol.qualifiers();
+                    List<Qualifier> typeDefQualifiers = typeDefinitionSymbol.qualifiers();
+                    if (hasQualifier(typeDefQualifiers, PUBLIC_KEYWORD) &&
+                            !hasQualifier(qualifiers, ISOLATED_KEYWORD)) {
+                        reportIssue(typeDefinitionNode, CoreRule.PUBLIC_NON_ISOLATED_CLASS_OR_OBJECT_CONSTRUCT);
+                    }
+                }
+            }
+        });
+        this.visitSyntaxNode(typeDefinitionNode);
+    }
+
+    private void checkNonIsolatedPublicClassDefinition(ClassDefinitionNode classDefinitionNode) {
+        semanticModel.symbol(classDefinitionNode).ifPresent(symbol -> {
+            if (symbol instanceof ObjectTypeSymbol objectTypeSymbol) {
+                List<Qualifier> qualifiers = objectTypeSymbol.qualifiers();
+                if (hasQualifier(qualifiers, PUBLIC_KEYWORD) &&
+                        !hasQualifier(qualifiers, ISOLATED_KEYWORD)) {
+                    reportIssue(classDefinitionNode, CoreRule.PUBLIC_NON_ISOLATED_CLASS_OR_OBJECT_CONSTRUCT);
+                }
+            }
+        });
+    }
+
+    private void checkNonIsolatedPublicClassMembers(ClassDefinitionNode classDefinitionNode) {
+        semanticModel.symbol(classDefinitionNode).ifPresent(classSymbol -> {
+            if (classSymbol instanceof ObjectTypeSymbol objectTypeSymbol) {
+                boolean isPublicObjectTypeSymbol = hasQualifier(objectTypeSymbol.qualifiers(), PUBLIC_KEYWORD);
+                classDefinitionNode.members().forEach(member -> {
+                    semanticModel.symbol(member).ifPresent(memberSymbol -> {
+                        if (isPublicObjectTypeSymbol && memberSymbol.kind() == SymbolKind.METHOD) {
+                            checkNonIsolatedPublicMethods((FunctionDefinitionNode) member);
+                        }
+                        member.accept(this);
```

Comment:
I think it should be part of the class visit, the method name here doesn't reflect this.